corehq/apps/data_interfaces/static/data_interfaces/js/case_dedupe_main.js

```diff
@@ -14,19 +14,156 @@ hqDefine("data_interfaces/js/case_dedupe_main", [
+                currentProperties = reserveProperties([property.name()], currentProperties);
+                rebuildPropertyMap(currentProperties);
+            }
+        };
+
+        self.removeProperty = function (property) {
+            self.selectedProperties.remove(property);
+
+            let currentProperties = self.availablePropertyMap()[self.caseType()];
+            currentProperties = restoreProperty(property.name(), currentProperties);
+
+            rebuildPropertyMap(currentProperties);
+        };
+
+        self.updatePropertyValue = function (oldValue, newValue) {
+            let currentProperties = self.availablePropertyMap()[self.caseType()];
+            currentProperties = reserveProperties([oldValue], currentProperties);
+            currentProperties = restoreProperty(newValue, currentProperties);
+
+            rebuildPropertyMap(currentProperties);
+        };
+
+        self.caseType.subscribe(handleCaseTypeChange);
+        function handleCaseTypeChange(caseType) {
+            if (!caseType) {
+                return;
+            }
+
+            resetAvailableProperties();
+
+            const selectedNames = self.selectedProperties().map(function (prop) {
+                return prop.name();
+            });
+
+            let currentProperties = self.availablePropertyMap()[caseType];
+            currentProperties = reserveProperties(selectedNames, currentProperties);
+            rebuildPropertyMap(currentProperties);
+        }
+
+        function resetAvailableProperties() {
+            const originalPropertyMap = clonePropertyMap(initialPropertyMap);
+            self.availablePropertyMap(originalPropertyMap);
+        }
+
+        function clonePropertyMap(originalMap) {
+            const clone = {};
+            for (const key in originalMap) {
+                clone[key] = originalMap[key].slice();
+            }
+
+            return clone;
+        }
+
+        function propertyExistsInOriginalMap(propName) {
+            const originalProperties = initialPropertyMap[self.caseType()];
+
+            return (originalProperties.indexOf(propName) !== -1);
+        }
+
+        function rebuildPropertyMap(changedCategory) {
+            const caseType = self.caseType();
+            const existingMap = self.availablePropertyMap();
+            changedCategory.sort();
+            existingMap[caseType] = changedCategory;
+            self.availablePropertyMap(existingMap);
+        }
+
+        function restoreProperty(propName, currentProperties) {
+            if (!propName || !propertyExistsInOriginalMap(propName)) {
+                // do not put custom properties back into the map
+                return currentProperties;
+            }
+
+            currentProperties.push(propName);
+
+            return currentProperties;
+        }
+
+        function reserveProperties(propertyNames, currentProperties) {
+            propertyNames.forEach(function (propName) {
```

Comment:
i vaguely remember `requirejs` having issues with `forEach` at some point way back, but maybe it's since been supported. It's why a lot of the code uses `uderscore`'s `_.each`. You can test by building `requirejs` locally (it will also fail on the `staging` build during deploy if unsupported)

Reply:
I did not encounter any errors building `requirejs` locally, so I assume this issue was fixed? I'm going to deploy this to staging regardless, so that will be further assurance.